lib/uefi/runtime_service_provider.cpp

```diff
@@ -61,6 +83,25 @@ EFI_STATUS SetVirtualAddressMap(size_t MemoryMapSize, size_t DescriptorSize,
 
 EFI_STATUS SetVariable(char16_t *VariableName, EfiGuid *VendorGuid,
                        uint32_t Attributes, size_t DataSize, void *Data) {
+  if (!VariableName || VariableName[0] == 0) {
+    return INVALID_PARAMETER;
+  }
+
+  /* Only allow setting non-volatile variables */
+  if ((Attributes & (~(EFI_VARIABLE_BOOTSERVICE_ACCESS | EFI_VARIABLE_RUNTIME_ACCESS))) == 0) {
+    char buffer[512];
+    memset(buffer, 0, sizeof(buffer));
+    for (size_t i = 0; i < sizeof(buffer) && VariableName[i] != 0; i++) {
+      buffer[i] = VariableName[i];
```

Comment:
this has happened enough times that I think we should add a `utf16_to_utf8` helper function, even if that helper function currently only handles simple ascii chars.